ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,2157 @@
+            ; mov r9, [r9 + 8]
+          ),
+        }
+        // Section 3.2.3 of the SysV AMD64 ABI:
+        // > The size of each argument gets rounded up to eightbytes. [...] Therefore the stack will always be eightbyte aligned.
+        self.offset_trampoline += 8;
+      }
+
+      (6.., param) => {
+        let ot = self.offset_trampoline as i32;
+        let oc = self.offset_callee as i32;
+        match param {
+          U(B) => x64!(s
+            // TODO: optimize to [rsp] (without immediate) when offset is 0
+            ; movzx eax, BYTE [rsp + ot]
+            ; mov [rsp + oc], eax
+          ),
+          I(B) => x64!(s
+            ; movsx eax, BYTE [rsp + ot]
+            ; mov [rsp + oc], eax
+          ),
+          U(W) => x64!(s
+            ; movzx eax, WORD [rsp + ot]
+            ; mov [rsp + oc], eax
+          ),
+          I(W) => x64!(s
+            ; movsx eax, WORD [rsp + ot]
+            ; mov [rsp + oc], eax
+          ),
+          U(DW) | I(DW) => x64!(s
+            ; mov eax, [rsp + ot]
+            ; mov [rsp + oc], eax
+          ),
+          U(QW) | I(QW) => x64!(s
+            ; mov rax, [rsp + ot]
+            ; mov [rsp + oc], rax
+          ),
+          Buffer => x64!(s
+            ; mov rax, [rsp + ot]
+            ; mov rax, [rax + 8]
+            ; mov [rsp + oc], rax
+          ),
+        }
+        // Section 3.2.3 of the SysV AMD64 ABI:
+        // > The size of each argument gets rounded up to eightbytes. [...] Therefore the stack will always be eightbyte aligned.
+        self.offset_trampoline += 8;
+        self.offset_callee += 8;
+
+        debug_assert!(
+          self.allocated_stack == 0
+            || self.offset_callee <= self.allocated_stack
+        );
+      }
+    }
+    self.integer_params += 1;
+  }
+
+  fn zero_first_arg(&mut self) {
+    debug_assert!(
+      self.integer_params == 0,
+      "the trampoline would zero the first argument after having overridden it with the second one"
+    );
+    dynasm!(self.assmblr
+      ; .arch x64
+      ; xor edi, edi
+    );
+  }
+
+  fn cast_return_value(&mut self, rv: NativeType) {
+    let s = &mut self.assmblr;
+    // V8 only supports 32bit integers. We support 8 and 16 bit integers casting them to 32bits.
+    // In SysV-AMD64 the convention dictates that the unused bits of the return value contain garbage, so we
+    // need to zero/sign extend the return value explicitly
+    match rv {
+      NativeType::U8 => x64!(s; movzx eax, al),
+      NativeType::I8 => x64!(s; movsx eax, al),
+      NativeType::U16 => x64!(s; movzx eax, ax),
+      NativeType::I16 => x64!(s; movsx eax, ax),
+      _ => (),
+    }
+  }
+
+  fn save_out_array_to_preserved_register(&mut self) {
+    let s = &mut self.assmblr;
+    // functions returning 64 bit integers have the out array appended as their last parameter,
+    // and it is a *FastApiTypedArray<Int32>
+    match self.integer_params {
+      // rdi is always V8 receiver
```

Comment:
rdi has been zeroed by the time this function is called, right?

Reply:
Yeah, not the most fortunate comment. I've elaborated it a bit, what do you think?